### ISSUE TITLE
Unified error console and error reporting

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#1090]
## 🚀 Feature
As discussed in meetings, napari should have an embedded, unified error console.  The idea being that, for the most "bundled" like usage scenarios, a user should not have to look in the console that started the napari process to see errors that occured.  It should also allow easy review and reporting of errors that have occured in the session, along the lines of #1024, but for all errors raised in napari. This would probably supersede #794 

**UPDATED 5/28 with Links to relevant topics:** 

- #794 describes a mechanism in which we would override `sys.excepthook` with our own handler.  That handler would have some logic that decides when to pop up an error to the user (for instance, if the exception was a subclass of some `NapariError` class ... or perhaps we would want to be even more permissive about what got shown).  This is probably the main thing we need to figure out and get right here.
- #1024 has some code that shows how to store and show tracebacks from errors that were previously raised.  Note: a key idea there was that `napari-plugin-engine` has a base exception class (`PluginError`) that [stores _all_ instances that are created in the class `_record` attribute](https://github.com/napari/napari-plugin-engine/blob/7ce84475aa003bdcd3e181b755be0e2e61fb14e8/napari_plugin_engine/exceptions.py#L14-L48).  We could do something similar here that would let the user review past exceptions, and potentially send them to us.
- see also #1303, that has the good idea to have this sort of "error report window" default to showing the most recent error...
- #1301 is another request to make certain errors (in that case, when opening an image fails) more visible to the user.  That's a key idea here, but rather than having a lot of `QDialog` instances scattered through the code, we should define a general API that we use throughout the codebase to show an error message in the GUI.  (for instance, as shown in #794)
- some of this stuff might be relevant for telemetry and bug reporting #621

**Comment by justinelarsen** [comment]
Yes, this would be helpful. 

For example, I dragged an image to the napari canvas, but nothing happened. I saw an error in the console “No plugin found capable of reading ... .tiff.” It would be nice to have a user friendly message that displayed within napari and not just the console.

**Comment by jni** [comment]
@justinelarsen well that is already a worrisome bug since we should definitely be able to open tiffs with the default plugin!!!! Did we forget to check for .tiff in addition to .tif?

**Comment by tlambert03** [comment]
we don't check file extensions at all ... that has to be an error in the `magic_imread` somehow.  @justinelarsen, can you send a link to the tiff you used?  Or after you get the error, go into the `Plugins > Plugin Errors...` menu, select "builtins" from the dropdown, and then click the button at the top right to copy the Traceback and paste it here?

**Comment by justinelarsen** [comment]
The tiff file used was an example file I downloaded from the openslide library test data: https://openslide.org/formats/
File is CMU-1.tiff: http://openslide.cs.cmu.edu/download/openslide-testdata/Generic-TIFF/

Traceback from the Plugin Errors menu below: (great feature by the way!)

```pytb
=========================== Errors for plugin 'builtins' ===========================

  napari version: 0.2.12+51.g6ea95a9


ERROR 1:  Error in plugin 'builtins', hook 'napari_get_reader' ------------------

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
~/dev/napari/napari/plugins/io.py in read_data_with_plugins(path='/Users/jlarsen/Downloads/CMU-1.tiff', plugin_manager=<napari.plugins.manager.PluginManager object>)
     53         try:
---> 54             return reader(path)  # try to read the data.
        reader = None
        path = '/Users/jlarsen/Downloads/CMU-1.tiff'
     55         except Exception as exc:

~/dev/napari/napari/types.py in reader_function(*args=('/Users/jlarsen/Downloads/CMU-1.tiff',), **kwargs={})
     48     def reader_function(*args, **kwargs) -> List[LayerData]:
---> 49         result = func(*args, **kwargs)
        result = undefined
        global func = undefined
        args = ('/Users/jlarsen/Downloads/CMU-1.tiff',)
        kwargs = {}
     50         return [(result,)]

~/dev/napari/napari/utils/io.py in magic_imread(filenames=['/Users/jlarsen/Downloads/CMU-1.tiff'], use_dask=False, stack=True)
    147             if shape is None:
--> 148                 image = imread(filename)
        image = undefined
        global imread = <function imread at 0x7fe6c8f7e3b0>
        filename = '/Users/jlarsen/Downloads/CMU-1.tiff'
    149                 shape = image.shape

~/dev/napari/napari/utils/io.py in imread(filename='/Users/jlarsen/Downloads/CMU-1.tiff')
     48     if ext in [".tif", ".tiff", ".lsm"]:
---> 49         import tifffile
        tifffile = undefined
     50 

ModuleNotFoundError: No module named 'tifffile'

The above exception was the direct cause of the following exception:

PluginError                               Traceback (most recent call last)
PluginError: Error in plugin 'builtins', hook 'napari_get_reader'
================================================================================
```

**Comment by tlambert03** [comment]
ah, looks like the same issue as https://github.com/napari/napari/issues/1129.
in that case, he was using the github master branch, but hadn't updated his requirements since #1061 merged.  this would not be a problem in the release, but in a developer environment it is.  so hopefully you just need to rerun `pip install -r requirements.txt -U`

update: yeah that file works for me as long as tifffile is installed

**Comment by justinelarsen** [comment]
Yes, that worked! Thanks, @tlambert03 !

**Comment by tlambert03** [comment]
just updated the origin post in this thread with links to various important discussions going on about how we catch, store, and show exceptions to users in the GUI.  which is going to be essential for a bundled app

**Comment by sofroniewn** [comment]
See also #1145 "Feedback requested on plugin exception handling."

**Comment by liu-ziyang** [comment]
@tlambert03 hey talley I am working on a proposal for this, I am still early and naive in my thoughts now, and would like to learn more about the issue and the existing discussions/ideas, as well as the actual scenarios where errors are raised so I have better perspectives how to handle them. Are you down for a meeting some time?

**Comment by liu-ziyang** [comment]
leaving my 2 cents here as well:
I am using chrome plugin development/debugging as a reference: https://developer.chrome.com/extensions/tut_debugging

overall the direction I am heading is:
1. In default mode, which is for the general users that do not have particular interest/skill to debug an issue, we can use a pop-up message to indicate that something went wrong, not sure if qt has this built-in but if not we can use some other library to achieve this. maybe have some reporting mechanism that posts the error back to core dev/plugin dev
2. similar to chrome, we can include a developer mode that has the console to deal with stack trace, etc, which needs to be enabled specifically by the developer to show the console interface. which would have more information like debugging logging, stack trace, etc. available

**Comment by tlambert03** [comment]
> Are you down for a meeting some time?

definitely, happy to chat about it.  I think the two points you mentioned here are along the same lines of what I was envisioning.

For point number 1, see #794.  The way to catch arbitrary exceptions without knowing ahead of time what they are going to be is to override `sys.excepthook` with our own [handler](https://github.com/napari/napari/pull/794/files#diff-6c077943a703ae4bebe19b6df0b91c8dR11).  That handler is then responsible for delivering the pop-up to the user, with something like one of the [static members of `QMessageBox`](https://doc.qt.io/qt-5/qmessagebox.html#static-public-members).  Exactly _what_ we show to the user in that popup is up for debate, but we can also create our own hierarchy of internal exception types that the handler is prepared to handle in a particular way (or just fall back to some generic "an error occurred" with the text of the error message for generic exception types).

For number 2, I wasn't necessarily envisioning that this would be a special dev mode.  While I don't think we should be showing stack traces right off the bat, i do think we should be capturing them and making them either available for viewing, or sending to the developer (i.e. napari issues tracker) with the click of a button.  Since we already have to gather all that stuff for a standard end-user to be able to send a stack trace, we might as well have a "show full error" button that anyone can click, without them having to reboot in a special mode.  Here, I'm envisioning something like the current `Plugin > Plugin Errors ...` menu (#1024, but with a more user-friendly veneer, and with the stack traces buried behind one more button).

(This is all assuming for the moment that we're not doing any telemetry #621 ... which of course might change some of our strategy.)

So to summarize:  the general goal is to make sure that the end-user never "wonders whether something happened or not".  We shouldn't need a console open to know that an exception occured.  but the first appearance of the error message should be friendly, with the option to dig deeper or send a full stack trace to napari or the appropriate plugin developer.

**Comment by sofroniewn** [comment]
> (This is all assuming for the moment that we're not doing any telemetry #621 ... which of course might change some of our strategy.)

I think we should be preparing for telemetry and designing systems now that could be used in such a way in an opt in system. I do like the post to GitHub too though, it's sort of less magical and mysterious then telemetry and maybe nice to have too. cc @neuromusic

**Comment by tlambert03** [comment]
it's not a hard switch... so I don't think it needs to affect this much at all.  The main thing that might change eventually is that we remove the button that says "send the stack trace", and perhaps remove some of the code that temporarily stores a stack trace for review.  But i don't think it will be a wasted effort regardless of how/whether we end up using sentry.

**Comment by neuromusic** [comment]
Even if automated error reporting gets implemented one day, you still need to think of how/if you support those users who opt-out of automated bug reporting. I suspect you'd still want to give them an easy way to send an individual bug report?

**Comment by liu-ziyang** [comment]
> The way to catch arbitrary exceptions without knowing ahead of time what they are going to be is to override `sys.excepthook` with our own [handler](https://github.com/napari/napari/pull/794/files#diff-6c077943a703ae4bebe19b6df0b91c8dR11).

I am with you on the arbitrary exception part that overriding the exception hook, though I am hoping that, we will unify the error handling through the defined procedures (the _on_error_xxxx call that I mentioned). I envision us promoting this method, which handles the error, including logging of the error using a logging library (instead of print() or w/e that people just do), sending reports, etc. What I am trying to achieve is a unified error handler that everyone calls when they need to work with exceptions. And yes, people can still choose to throw the exception all the way up, which we can work around by also overriding the exception hook to our error handler

**Comment by tlambert03** [comment]
where do you see this `on_error` call living?  and how exactly is it different from the error handler that I'm talking about (the one that we use to override `sys.excepthook`.  When you say "promoting this method" what do you mean by that?  With `sys.excepthook`, we wouldn't be promoting a new way of raiding errors.  People just use pure python (`raise Exception`) from wherever they'd like, and we handle it with our handler.  You can certainly use standard logging (definitely not print!) in the handler if desired.

>  What I am trying to achieve is a unified error handler that everyone calls when they need to work with exceptions

same here... I'm struggling to understand how you can get more "unified" that overriding the root python hook that handles exceptions.  I _don't_ think we should be promoting a new way of calling errors.  I think we should stick with the standard `raise Exception()`... (though perhaps we might define `raise NapariErrorType()` as mentioned before).   Maybe if you can provide an example bit of code and where you see it living, it might help me to understand what you're proposing, and how it differs from the pattern in #794

**Comment by liu-ziyang** [comment]
sorry I think I have been staying too abstract and high level, let me try again with some code snippet, maybe this makes more sense to what I am trying to achieve:

in a typical try block, you may have:
```
try:
    xyz()
except e:
    print(e)
```
I would want to include these use cases to our error handler, instead of a highest level catch all method, we unify exception handling logic and promote this method to developers, so instead we would allow user to use our error handler and write this instead (syntax yet to be defined, we can use annotation or w/e to achieve this):
```
on_error_ignore(xyz) 
```

**Comment by tlambert03** [comment]
I apologize, but I'm afraid I'm still not following your vision here.  by "provide an example bit of code", I meant a little more than what you show here.  For instance, _where_ are you calling your `on_error_ignore` function here?  one place?  in every except clause? in place of a try/except block? what does your proposed `on_error_ignore` function _do_?  if you think it would be helpful to draft up a PR with some of your ideas, I'd be happy to look at it.  

so far though, these would be some of my questions:

- do you see what you're proposing as an _alternative_ to a `try/except/raise` and `sys.excepthook`-based error handler?  or more like "yes, and eventually someday we might also add something like `on_error_ignore()`"
- if the former, then what do you see as the advantage of adding a new syntax and pattern that developers should learn and use, rather than just using `raise Exception()` in the standard way?  in other words, what is the standard `try/catch/raise` pattern lacking for you?
- for the main task at hand (specifically: making sure users always know that/why an error was raised, and that we have a way to capture/report tracebacks), what limitations did you specifically see in the  `sys.excepthook`/handler pattern, and how does what you're proposing solve them?
- when you say "allow user to use our error handler", can you elaborate?
- was it completely clear how the code in #794 works?  Is it clear that a developer could already "use" our handler simply by doing a standard try/except like this:
    ```python
    try:
        xyz()
    except:
        raise Exception("something wrong")
    ```
   and that _all_ exceptions like that go to our handler that overrides `sys.excepthook`, where we can then alert the user _instead_ of having a message printed to stdout in the terminal (which a GUI user may or may not be looking at) unless of course, we _chose_ to print the traceback to the terminal in, for instance, an interactive session?.  and is it clear that we could additionally declare specific exception types (e.g. `raise NapariError())` that would have specific implications for how they are handled?  ... and that any developer could raise our error classes if they wanted to?

**Comment by liu-ziyang** [comment]
let me use this issue as an example to demonstrate what I meant: https://github.com/napari/napari/issues/1100
The napari code base rarely does anything with internal error/exception, almost everything is thrown all the way up the stack to the global level, which I do not feel appropriate.
to answer your question: no I am not trying to promote an alternative to try-catch, but an alternative to the default coding mindset that does not use much try-catch methodology at all. 

Just to be clear, I do not disagree with the principle of change in #794, in fact, I love the behavior how the error message is displayed, as well as the way to click a button to see details, I found that design aligns very well with overall software usage patterns. What I am planning to cover in the error reporting/handling proposal is overall hardening of the flow with more and easier exception handling. For example, if I added this one line at the end of add_labels.py example:
```
    label_layer.selected_label = -1
```
then the whole program would crash.  I do not believe crashing is a right behavior. Sure, by all means, it is a user error. But this could as well happen from internal code where underlying libraries raise runtime errors. 
I want to center my point around: I want to harden the overall error handling logic, to provide a way where we can insert more intermediate error control in the flow, where we made it easy and provided a consistent pattern/function for. To a point where there should rarely be a case where we need to pop up a message telling people something went wrong. But still have all of the unexpected events logged, and properly defaulted/ignored/raised

**Comment by tlambert03** [comment]
> I do not believe crashing is a right behavior. 

ooooh... is that the thing you're trying to avoid? I agree 100%: crashing is not a good thing! 😂. This is actually a qt thing.  Since pyqt5.5, an unhandled Python exception will result in a call to Qt’s `qFatal()` function, which is what you're seeing there.  See [this stack overflow post](https://stackoverflow.com/questions/33736819/pyqt-no-error-msg-traceback-on-exit/33741755#33741755) for more background on that.

The solution to that is actually #794!  In fact, wanting to solve that exact problem is how I learned about this pattern in a [different qt-based program I once worked on](https://github.com/tlambert03/LLSpy/blob/develop/llspy/gui/exceptions.py).  If I add back the logic from #794 and add `label_layer.selected_label = -1` to the end of that example, i get a nice friendly popup and the program does _not_ crash:

<img width="424" alt="Screen Shot 2020-06-21 at 6 15 39 AM" src="https://user-images.githubusercontent.com/1609449/85222206-64b8a200-b387-11ea-916a-ba8fd91732c3.png">

Now, I _do_ agree it's a good idea to generally promote more `try/except` in our code, and the very important thing here (in my opinion) is that the exceptions raised in those blocks should have extremely clear error messages.  However, you're _never_ going to get to a point where you catch them all.  #794 PR was _not_ actually about the how the error message looked or the fact that you can click a button to see details.  It was about handling _uncaught_ exceptions, (the type we didn't anticipate) in a graceful manner such it's a) clear that an error occurred and b) that the program doesn't crash in scripts.

Getting to a point where we never have to pop up an error is great, it definitely means that we've anticipated the majority of things that can go wrong.  But I'm not sure what that PR "looks like".  That comes with development of a program, as you find problem areas and wrap them in try/except blocks.  This proposal was about how to handle the inevitable fact that we will never actually get to a point where we don't have uncaught exceptions.

**Comment by liu-ziyang** [comment]
Ohhh, I think there might be some misunderstanding of what I am going to produce. I am working on a proposal for the overall design or error handling, instead of actual code change, which likely won't be addressed in a single PR. The proposal will likely start with the imminent global error exception handler override, to address the issue where users can't see what went wrong and will expand gradually into the future to enhance the code overall. Let me write more things down so people can take a look, I think I did not do a great job explaining what I had in mind, and the explanation that I gave is too high level and abstract. I will get some more detailed analyzation written down in the coming week

**Comment by sofroniewn** [comment]
Just weighing in to say this discussion has been great so far, I'm learning a lot. I think both - how we handle unexpected errors and how to promote more `try`/ `catch` within the codebase and good handling of those errors we intentionally raise seem valuable!!

**Comment by liu-ziyang** [comment]
the design is ready to be reviewed: https://docs.google.com/document/d/1Bdpqm6WOvnBq00QhxnUFPHe55BzH4dZFrCs2CtNU_GA/edit#heading=h.lcy76pz8vp1k

I am not sure if I should open up another issue to post the content, it seems easier to comment in place on gdoc

**Comment by sofroniewn** [comment]
> I am not sure if I should open up another issue to post the content, it seems easier to comment in place on gdoc

I'd open a brand new issue and post content in there, will probably be easier to have discussion there. If that doesn't work we can revisit the gdoc approach (there I imagine there will be issues with permissions/ accounts etc)

**Comment by liu-ziyang** [comment]
sure, see: https://github.com/napari/napari/issues/1395

**Comment by pwinston** [comment]
Not sure if this is a good idea, but for perfmon we swap in our own `QApplication` which has its own version of `notify()`:
https://github.com/napari/napari/blob/master/napari/_qt/qt_event_timing.py#L60-L66

We do that so we can time every single Qt Event as it gets processed. If we put a try/except around that notify call, that we "protect" basically any code that results from any Qt event. I think this would include vispy events. Basically if the except happened it would mean that processing the event failed.

Maybe there is some downside from having a try/except there. But if we want to experiment with a very high level try/except this might be a place to do it. You could experiment just be setting NAPARI_PERFMON=1 and modifying our `QApplicationWithTiming`. If we wanted to do this long term we could replace `QApplicationWithTiming` with just `QApplicationNapari` or something and always insert it.
